File: systems-for-design-and-implementation/labs/exam/TicTacToeNetworking/src/tictactoe/network/rpcprotocol/Request.java
package tictactoe.network.rpcprotocol;

import java.io.Serializable;

public class Request implements Serializable
{
    private RequestType type;
    private Object      data;

    private Request()
    {
    }

    public RequestType getType()
    {
        return type;
    }

    private void setType(RequestType type)
    {
        this.type = type;
    }

    public Object getData()
    {
        return data;
    }

    private void setData(Object data)
    {
        this.data = data;
    }

    @Override
    public String toString()
    {
        return "Request{" +
            "getType='" + type + '\'' +
            ", getData='" + data + '\'' +
            '}';
    }

    public static class Builder
    {
        private Request request = new Request();

        public Builder type(RequestType type)
        {
            request.setType(type);
            return this;
        }

        public Builder data(Object data)
        {
            request.setData(data);
            return this;
        }

        public Request build()
        {
            return request;
        }
    }
}

File: systems-for-design-and-implementation/labs/lab3/LibraryNetworking/src/library/network/rpcprotocol/Request.scala
package library.network.rpcprotocol

import java.io.Serializable

object Request
{
    class Builder
    {
        private val request: Request = new Request

        def `type`(`type`: String): Request.Builder =
        {
            request.setType(`type`)
            this
        }

        def data(data: Any): Request.Builder =
        {
            request.setData(data)
            this
        }

        def build: Request =
        {
            request
        }
    }
}

class Request private extends Serializable
{
    private var `type`: String = null
    private var data: Any = null

    def getType: String =
    {
        `type`
    }

    private def setType(`type`: String)
    {
        this.`type` = `type`
    }

    def getData: Any =
    {
        data
    }

    private def setData(data: Any)
    {
        this.data = data
    }

    override def toString: String =
    {
        "Request{" + "getType='" + `type` + '\'' + ", getData='" + data + '\'' + '}'
    }
}
